File: vst3cwrapper/src/main/cpp/funknown.cpp
#include "FUnknown.h"

#include <pluginterfaces/base/funknown.h>

Steinberg::FUnknown *cast(FUnknown *this_ptr) {  // private
    return reinterpret_cast<Steinberg::FUnknown *>(this_ptr);
}

tresult FUnknown_queryInterface(FUnknown *this_ptr, TUID _iid, void **obj) {
    return cast(this_ptr)->queryInterface(_iid, obj);
}

uint32 FUnknown_addRef(FUnknown *this_ptr) {
    return cast(this_ptr)->addRef();
}

uint32 FUnknown_release(FUnknown *this_ptr) {
    return cast(this_ptr)->release();
}

File: interop/src/linuxX64Main/kotlin/io/github/durun/vst3kotlin/base/FUnknown.kt
package io.github.durun.vst3kotlin.base

import cwrapper.*
import cwrapper.FUnknown
import io.github.durun.io.Closeable
import io.github.durun.vst3kotlin.VstInterface
import kotlinx.cinterop.*

actual abstract class FUnknown(
	protected val thisRawPtr: CPointer<*>
) : Closeable {
	private val ptr: CPointer<FUnknown> get() = thisRawPtr.reinterpret()
	actual final override var isOpen: Boolean = true
		private set

	actual override fun close() {
		check(isOpen)
		FUnknown_release(ptr)
		isOpen = false
	}

	fun queryInterface(interfaceID: TUID): CPointer<*> {
		return memScoped {
			val obj = alloc<CPointerVar<*>>()
			val result = IPluginFactory_queryInterface( // TODO: use FUnknown_queryInterface
				ptr.reinterpret(),
				interfaceID,
				obj.ptr
			)
			check(result == kResultTrue) { result.kResultString }
			val ptr = obj.value
			checkNotNull(ptr)
			ptr
		}
	}

	inline fun <reified I : CPointed> queryInterface(): VstInterface<CPointer<I>> {
		val iid = when (I::class) {
			IComponent::class -> IComponent_iid
			IComponentHandler::class -> IComponentHandler_iid
			IComponentHandler2::class -> IComponentHandler2_iid
			IEditController::class -> IEditController_iid
			IEditController2::class -> IEditController2_iid
			IProcessContextRequirements::class -> IProcessContextRequirements_iid
			else -> throw IllegalArgumentException()
		}
		val ptr = queryInterface(iid).reinterpret<I>()
		return VstInterface(ptr)
	}
}